File: tools/sec_filings_tool.py
"""SEC EDGAR Filing Retrieval.

Fetches 10-K annual reports for public companies. No API key required.
SEC fair-access policy requires a descriptive User-Agent header.
"""

from __future__ import annotations

import json
from typing import Any

import httpx
from claude_agent_sdk import tool

SEC_HEADERS = {
    "User-Agent": "LaszyResearch/1.0 (albert@zip.co)",
    "Accept-Encoding": "gzip, deflate",
}


@tool(
    "fetch_sec_filings",
    "Fetch the most recent 10-K filing from SEC EDGAR for a public company. Returns filing text content (capped at ~100K chars). For private companies, returns an empty result with a note.",
    {"company_name": str, "ticker": str},
)
async def fetch_sec_filings(args: dict[str, Any]) -> dict[str, Any]:
    company_name = args["company_name"]
    ticker = args.get("ticker") or ""
    query = ticker if ticker and ticker != "unknown" else company_name

    try:
        async with httpx.AsyncClient(timeout=60.0, headers=SEC_HEADERS) as client:
            search_resp = await client.get(
                "https://efts.sec.gov/LATEST/search-index",
                params={
                    "q": query,
                    "dateRange": "custom",
                    "startdt": "2024-01-01",
                    "forms": "10-K",
                },
            )
            search_resp.raise_for_status()
            results = search_resp.json()

            hits = results.get("hits", {}).get("hits", [])
            if not hits:
                return _content({
                    "error": f"No 10-K found for {company_name}. Company may be private.",
                    "filings": [],
                    "company": company_name,
                })

            filing_source = hits[0]["_source"]
            filing_url = filing_source.get("file_url", "")
            if not filing_url:
                return _content({
                    "error": "Filing URL not found in search results",
                    "filings": [],
                    "company": company_name,
                })

            filing_resp = await client.get(f"https://www.sec.gov{filing_url}")
            filing_resp.raise_for_status()
            content = filing_resp.text[:100_000]

        return _content({
            "filing_type": "10-K",
            "company": company_name,
            "content": content,
            "source_url": f"https://www.sec.gov{filing_url}",
            "filing_date": filing_source.get("file_date", "unknown"),
        })

    except httpx.HTTPError as e:
        return _content({
            "error": f"SEC EDGAR request failed: {e}",
            "filings": [],
            "company": company_name,
        })


def _content(payload: dict[str, Any]) -> dict[str, Any]:
    return {"content": [{"type": "text", "text": json.dumps(payload)}]}

File: src/tools/secFilingsTool.ts
import { OpenAITool } from "./openaiTools.js";

const SEC_HEADERS = {
  "User-Agent": "LaszyResearch/1.0 (albert@zip.co)",
  "Accept-Encoding": "gzip, deflate"
};

export async function fetchSecFilings(
  companyName: string,
  ticker?: string
): Promise<Record<string, unknown>> {
  const query = ticker && ticker !== "unknown" ? ticker : companyName;

  try {
    const searchUrl = new URL("https://efts.sec.gov/LATEST/search-index");
    searchUrl.searchParams.set("q", query);
    searchUrl.searchParams.set("dateRange", "custom");
    searchUrl.searchParams.set("startdt", "2024-01-01");
    searchUrl.searchParams.set("forms", "10-K");

    const searchResponse = await fetch(searchUrl, {
      headers: SEC_HEADERS,
      signal: AbortSignal.timeout(30_000)
    });

    if (!searchResponse.ok) {
      return {
        error: `SEC EDGAR request failed: ${searchResponse.status} ${searchResponse.statusText}`,
        filings: [],
        company: companyName
      };
    }

    const results = (await searchResponse.json()) as {
      hits?: { hits?: Array<{ _source?: Record<string, unknown> }> };
    };
    const hits = results.hits?.hits ?? [];
    if (hits.length === 0) {
      return {
        error: `No 10-K found for ${companyName}. Company may be private.`,
        filings: [],
        company: companyName
      };
    }

    const filingSource = hits[0]._source ?? {};
    const filingUrl = String(filingSource.file_url ?? "");
    if (!filingUrl) {
      return {
        error: "Filing URL not found in search results",
        filings: [],
        company: companyName
      };
    }

    const filingResponse = await fetch(`https://www.sec.gov${filingUrl}`, {
      headers: SEC_HEADERS,
      signal: AbortSignal.timeout(60_000)
    });

    if (!filingResponse.ok) {
      return {
        error: `SEC filing fetch failed: ${filingResponse.status} ${filingResponse.statusText}`,
        filings: [],
        company: companyName
      };
    }

    const content = (await filingResponse.text()).slice(0, 100_000);

    return {
      filing_type: "10-K",
      company: companyName,
      content,
      source_url: `https://www.sec.gov${filingUrl}`,
      filing_date: filingSource.file_date ?? "unknown"
    };
  } catch (error) {
    return {
      error: `SEC EDGAR request failed: ${error instanceof Error ? error.message : String(error)}`,
      filings: [],
      company: companyName
    };
  }
}

export const fetchSecFilingsTool = new OpenAITool({
  name: "fetch_sec_filings",
  description: "Fetch the most recent SEC 10-K filing for a public company.",
  parameters: {
    type: "object",
    properties: {
      company_name: { type: "string" },
      ticker: { type: "string" }
    },
    required: ["company_name"],
    additionalProperties: false
  },
  handler: (args) =>
    fetchSecFilings(
      String(args.company_name ?? ""),
      args.ticker === undefined ? undefined : String(args.ticker)
    )
});
